qemu/tests/win_virtio_driver_installer_uninstall.py

```diff
@@ -62,42 +62,56 @@ def run(test, params, env):
                          expected_gagent_version)
     driver_check(session, test, params)
 
-    error_context.context("Run virtio-win-guest-tools.exe uninstall test",
-                          test.log.info)
-    vm.send_key('meta_l-d')
-    time.sleep(30)
-    run_uninstall_cmd = utils_misc.set_winutils_letter(
-        session, params["run_uninstall_cmd"])
-    session.cmd(run_uninstall_cmd)
-    time.sleep(30)
-    s_check, o_check = session.cmd_status_output(installer_pkg_check_cmd)
-    if s_check == 0:
-        test.fail("Could not uninstall Virtio-win-guest-tools package "
-                  "in guest', detail: '%s'" % o_check)
-    session = vm.reboot(session)
+    if params.get_boolean("uninstall_by_msi", False):
+        error_context.context("Check the warning when use msi to uninstall driver",
+                              test.log.info)
+        vm.send_key('meta_l-d')
+        time.sleep(30)
+        run_uninstall_cmd = utils_misc.set_winutils_letter(
+            session, params["run_uninstall_cmd"])
+        session.cmd(run_uninstall_cmd)
+        time.sleep(30)
+        check_warning_file = params["check_warning_file"]
+        output = session.cmd_output(check_warning_file)
+        if params["warning_message"] not in output:
+            test.fail("Not found expected warning message, the output is %s" % output)
+    else:
+        error_context.context("Run virtio-win-guest-tools.exe uninstall test",
+                              test.log.info)
+        vm.send_key('meta_l-d')
+        time.sleep(30)
+        run_uninstall_cmd = utils_misc.set_winutils_letter(
+            session, params["run_uninstall_cmd"])
+        session.cmd(run_uninstall_cmd)
+        time.sleep(30)
+        s_check, o_check = session.cmd_status_output(installer_pkg_check_cmd)
+        if s_check == 0:
+            test.fail("Could not uninstall Virtio-win-guest-tools package "
+                      "in guest', detail: '%s'" % o_check)
+        session = vm.reboot(session)
```

Comment:
Let's introduce a params `uninstall_method`:

```suggestion
    uninstall_method = params.get('uninstall_method', 'installer')
    error_context.context("Run uninstall test via %s" % uninstall_method,
                              test.log.info)
    vm.send_key('meta_l-d')
    time.sleep(30)
    run_uninstall_cmd = utils_misc.set_winutils_letter(
        session, params["run_uninstall_cmd"])
    session.cmd(run_uninstall_cmd)
    time.sleep(30)
    if uninstall_method == "MSI":
        check_warning_file = params["check_warning_file"]
        output = session.cmd_output(check_warning_file)
        if params["warning_message"] not in output:
            test.fail("Not found expected warning message, the output is %s" % output)
    else:
        s_check, o_check = session.cmd_status_output(installer_pkg_check_cmd)
        if s_check == 0:
            test.fail("Could not uninstall Virtio-win-guest-tools package "
                      "in guest', detail: '%s'" % o_check)
        session = vm.reboot(session)
```

Reply:
Another question to the test step itself: so we totally ignore whether the uninstall is successful or not via MSI? @xiagao @heywji ?

Reply:
> Let's introduce a params `uninstall_method`:

Updated,thanks!

Reply:
Hi @vivianQizhu, Let me explain our test steps. The development has completely fixed the problem with missing DNS and IP addresses, but there are still unknown issues with some other parts of the virtio-win module. The developer still keeps this warning message, so our test step is to check that msi has this warning message when uninstalling. We do not ensure msi uninstalled is successful or not.

Reply:
@heywji I understand you don't need to check it in your test with netkvm, but let's confirm with @xiagao If we really want to ignore it.

Reply:
In my opinion, 'Checking warning info' is just a step in win_virtio_driver_install_by_installer/uninstall test case, no need to check the uninstall result as from the test steps, 'tips exist and you can close the window if you can see it.'